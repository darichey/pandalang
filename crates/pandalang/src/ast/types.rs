// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
