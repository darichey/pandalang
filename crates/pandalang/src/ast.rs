#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub patt: Pattern,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub patt: Pattern,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Pattern {
    Id {
        name: String,
        // typ: Type
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Base { name: String },
    Fun(Box<Type>, Box<Type>),
}
