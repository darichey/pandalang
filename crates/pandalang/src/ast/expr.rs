#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
