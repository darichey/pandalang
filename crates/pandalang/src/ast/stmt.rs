use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
