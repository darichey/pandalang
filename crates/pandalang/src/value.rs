use crate::{ast, eval::Env};

#[derive(Clone)]
pub enum Value {
    Int(ast::Int),
    Str(ast::Str),
    Fun { fun: ast::Fun, env: Env },
}

impl std::fmt::Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Value::Int(ast::Int { n }) => write!(f, "{}", n),
            Value::Str(ast::Str { s }) => write!(f, "{}", s),
            Value::Fun { .. } => write!(f, "<function>"),
        }
    }
}

impl std::fmt::Debug for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Int(n) => f.debug_tuple("Int").field(n).finish(),
            Self::Str(s) => f.debug_tuple("Str").field(s).finish(),
            Self::Fun { fun, env: _ } => f
                .debug_struct("Fun")
                .field("fun", fun)
                .field("env", &"<opaque>".to_string())
                .finish(),
        }
    }
}
