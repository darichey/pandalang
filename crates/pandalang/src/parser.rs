use lalrpop_util::{lexer::Token, ParseError};

use crate::ast::Expr;

lalrpop_mod!(
    #[allow(clippy::all)]
    #[allow(unused)]
    pub grammar
);

pub fn parse(s: &str) -> Result<Box<Expr>, ParseError<usize, Token<'_>, &'static str>> {
    grammar::ExprParser::new().parse(s)
}

#[cfg(test)]
mod tests {
    use std::path::Path;

    use crate::{ast::Expr, parser};

    fn test(path: &Path) -> Result<Box<Expr>, String> {
        let source = std::fs::read_to_string(path).map_err(|err| err.to_string())?;
        parser::parse(&source).map_err(|err| err.to_string())
    }

    #[test]
    fn parses() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            insta::assert_debug_snapshot!(test(path));
        });
    }
}
