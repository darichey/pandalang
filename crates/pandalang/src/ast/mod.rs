use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
