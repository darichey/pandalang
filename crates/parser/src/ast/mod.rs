use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
