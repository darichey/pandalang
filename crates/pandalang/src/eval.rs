use std::collections::HashMap;

use crate::ast::{App, BinOp, BinOpKind, Expr, Fun, Int, Var};
use crate::value::Value;

macro_rules! bindings {
    ($($k:expr => $v:expr),* $(,)?) => {{
        std::collections::HashMap::from([$(($k.to_string(), vec![Value::Int(Int { n: $v })]),)*])
    }};
}

#[derive(Debug, Clone)]
pub struct Env {
    pub bindings: HashMap<String, Vec<Value>>,
}

impl Env {
    pub fn new() -> Env {
        Env {
            bindings: bindings!(),
        }
    }

    pub fn eval(&mut self, expr: Expr) -> Value {
        match expr {
            Expr::Int(n) => Value::Int(n),
            Expr::Str(s) => Value::Str(s),
            Expr::Var(Var { name }) => self
                .lookup(&name)
                .unwrap_or_else(|| panic!("{} is not bound!", name)),
            Expr::BinOp(BinOp { left, right, kind }) => {
                let f = match kind {
                    BinOpKind::Add => std::ops::Add::add,
                    BinOpKind::Sub => std::ops::Sub::sub,
                    BinOpKind::Mul => std::ops::Mul::mul,
                    BinOpKind::Div => std::ops::Div::div,
                };

                let (x, y) = match (self.eval(*left), self.eval(*right)) {
                    (Value::Int(Int { n: x }), Value::Int(Int { n: y })) => (x, y),
                    _ => panic!("Cannot eval BinOp with non-Int operands"),
                };

                Value::Int(Int { n: f(x, y) })
            }
            Expr::Fun(fun) => Value::Fun {
                fun,
                env: self.clone(),
            },
            Expr::App(App { fun, arg }) => {
                let (arg_name, body, mut fun_env) = match self.eval(*fun) {
                    Value::Fun {
                        fun: Fun { arg, body },
                        env,
                    } => (arg, body, env),
                    _ => panic!("Cannot apply non-functions"),
                };
                let arg = self.eval(*arg);

                fun_env.push_binding(&arg_name, arg);
                let result = fun_env.eval(*body);
                fun_env.pop_binding(&arg_name);
                result
            }
            Expr::Let(_) => panic!("Let in eval"),
        }
    }

    fn lookup(&self, name: &String) -> Option<Value> {
        let bindings = self.bindings.get(name)?;
        println!("{} = {:?}", name, bindings);
        let value = bindings.last()?;
        Some(value.clone()) // TODO: story around cloning here?
    }

    fn push_binding(&mut self, name: &String, value: Value) {
        println!("push {} = {:?}", name, value);
        match self.bindings.get_mut(name) {
            Some(current_bindings) => current_bindings.push(value),
            None => {
                self.bindings.insert(name.clone(), vec![value]);
            }
        };
    }

    fn pop_binding(&mut self, name: &String) {
        println!("pop {}", name);
        if let Some(current_bindings) = self.bindings.get_mut(name) {
            current_bindings.pop();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Env;
    use crate::ast::Int;
    use crate::parser;
    use crate::value::Value;

    fn eval_test(s: String) -> Value {
        let mut env = Env {
            bindings: bindings!("x" => 0, "y" => 1, "x'" => 2, "foo" => 3, "a" => 4, "b" => 5, "c" => 6, "d" => 7, "e" => 8),
        };
        env.eval(*parser::parse(s.as_str()).unwrap())
    }

    #[test]
    fn evals() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            let source = eval_test(std::fs::read_to_string(path).unwrap());
            insta::assert_debug_snapshot!(source);
        });
    }
}
