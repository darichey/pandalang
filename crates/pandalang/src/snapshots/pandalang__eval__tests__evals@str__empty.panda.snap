---
source: crates/pandalang/src/eval.rs
expression: test(path)
input_file: crates/pandalang/src/snapshot_inputs/str/empty.panda
---
Ok(
    Str(
        Str {
            s: "",
        },
    ),
)
